File: TypeScript/merge_two_sorted_lists/21.test.ts
import { ListNode, mergeTwoLists } from './21';
import { describe, test, expect } from 'bun:test';

interface tt {
  list1: ListNode | null;
  list2: ListNode | null;
  want: ListNode | null;
}
const tests: tt[] = [
  {
    list1: new ListNode(1, new ListNode(2, new ListNode(4))),
    list2: new ListNode(1, new ListNode(3, new ListNode(4, null))),
    want: new ListNode(1, new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(4, null)))))),
  },
];

describe('merge two sorted lists', () => {
  tests.forEach((t, i) => {
    test(`case ${i + 1}`, () => {
      const result = mergeTwoLists(t.list1, t.list2);
      expect(result).toEqual(t.want);
    });
  });
});

File: Golang/merge_two_sorted_lists/21_test.go
package mergetwosortedlists

import (
	"fmt"
	"testing"

	helpers "leetcode"

	"github.com/stretchr/testify/assert"
)

func TestMergeTwoLists(t *testing.T) {
	cases := []struct {
		list1 ListNode
		list2 ListNode
		want  ListNode
	}{
		{
			list1: helpers.CreateList([]int{1, 2, 4}),
			list2: helpers.CreateList([]int{1, 3, 4}),
			want:  helpers.CreateList([]int{1, 1, 2, 3, 4, 4}),
		},
		{
			list1: helpers.CreateList([]int{2, 3, 5}),
			list2: helpers.CreateList([]int{2, 3, 9}),
			want:  helpers.CreateList([]int{2, 2, 3, 3, 5, 9}),
		},
	}

	for i, tt := range cases {
		t.Run(fmt.Sprintf("case %d", i+1), func(t *testing.T) {
			result := mergeTwoLists(&tt.list1, &tt.list2)
			assert.Equal(t, tt.want, *result)
		})
	}
}
